### PR TITLE
Added Position and Velocity Covariances, Added option to not transform obstacles into global frame

### PR DESCRIPTION
**Pull request by alexanderjyuen**
`Obstacle` Message has been updated to include the position and velocity covariance

The `ObstacleClass` also now saves the position and velocity covariance in the correction step

The position and velocity of the tracked obstacles are also updated in the prediction step in the same way they are updated in the correction step. This is done to ensure an obstacle does not stay stationary if it misses detections over multiple frames

Additional parameter `transform_to_global_frame` is included to give user the option of keeping the obstacle array in the sensor frame or transforming it into the global frame

### CODE REVIEW

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 1]
```diff
@@ -177,49 +200,55 @@ def callback(self, msg):
...
-                marker.type = 1 # CUBE
+                marker.type = 1  # CUBE
                 marker.action = 0
                 marker.color.a = 0.5
                 marker.color.r = r
                 marker.color.g = g
                 marker.color.b = b
                 marker.pose.position = obs.msg.position
                 angle = np.arctan2(obs.msg.velocity.y, obs.msg.velocity.x)
-                marker.pose.orientation.z = np.float(np.sin(angle / 2))
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py:194]: Why remove the orientation?

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 2]
```diff
@@ -93,38 +103,43 @@ def callback(self, msg):
...
-                    detections[i].size = do_transform_vector3(s, trans).vector
-
-            except TransformException as ex:
-                self.get_logger().error(
-                    'fail to get tf from {} to {}: {}'.format(
-                    msg.header.frame_id, self.global_frame, ex))
-                return
+        if self.transform_to_global_frame:
+            # transform to global frame
+            if self.global_frame is not None:
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: How about `if self.transform_to_global_frame && if self.global_frame is not None`? I don't think this needs the double nest

File: nav2_dynamic_msgs/msg/Obstacle.msg [hunk 3]
```diff
@@ -3,3 +3,5 @@ float32 score # detection confidence
 geometry_msgs/Point position # center
 geometry_msgs/Vector3 velocity
 geometry_msgs/Vector3 size
+geometry_msgs/Vector3 position_covariance
```
Comment (SteveMacenski) [inline nav2_dynamic_msgs/msg/Obstacle.msg]: I think these are usually represented by a dense matrix, no?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py (hunk 1 above)
Comment (alexanderjyuen) [reply kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py:194]: fixed in 6a7ebcf7a741774daf8a7306cc4b31b58aaf7412

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py (hunk 2 above)
Comment (alexanderjyuen) [reply kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: Fixed in 0dc3249974b9bdb869c41817edcfa4ea19c87503

File: nav2_dynamic_msgs/msg/Obstacle.msg (hunk 3 above)
Comment (alexanderjyuen) [reply nav2_dynamic_msgs/msg/Obstacle.msg]: The radar that I'm using presently only gives me RMS in only the the 3 linear degrees of freedom. However it wouldn't be difficult to make this adaptable for radars that have cross terms in the covariance matrix. Should I use a `float64[9]` for this?

File: kf_hungarian_tracker/kf_hungarian_tracker/obstacle_class.py [hunk 4]
```diff
@@ -15,92 +16,181 @@ class ObstacleClass:
...
-        dt3 = dt*dt2
+        dt3 = dt * dt2
         dt4 = dt2**2
-        Q = np.array([[dt4*self.process_noise_cov[0]/4, 0, 0, dt3*self.process_noise_cov[0]/2, 0, 0], 
-                      [0, dt4*self.process_noise_cov[1]/4, 0, 0, dt3*self.process_noise_cov[1]/2, 0],
-                      [0, 0, dt4*self.process_noise_cov[2]/4, 0, 0, dt3*self.process_noise_cov[2]/2],
-                      [dt3*self.process_noise_cov[0]/2, 0, 0, dt2*self.process_noise_cov[0], 0, 0],
-                      [0, dt3*self.process_noise_cov[1]/2, 0, 0, dt2*self.process_noise_cov[1], 0],
-                      [0, 0, dt3*self.process_noise_cov[2]/2, 0, 0, dt2*self.process_noise_cov[2]]]).astype(np.float32)
+        Q = np.array(
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/obstacle_class.py]: Please revert this, this is not matrix readable

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 5]
```diff
@@ -28,25 +31,28 @@ class KFHungarianTracker(Node):
...
+                ("global_frame", "camera_link"),
+                ("process_noise_cov", [2.0, 2.0, 0.5]),
+                ("top_down", False),
+                ("death_threshold", 3),
+                ("measurement_noise_cov", [1.0, 1.0, 1.0]),
+                ("error_cov_post", [1.0, 1.0, 1.0, 10.0, 10.0, 10.0]),
+                ("vel_filter", [0.1, 2.0]),
+                ("height_filter", [-2.0, 2.0]),
+                ("cost_filter", 1.0),
+                ("transform_to_global_frame", False),
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: Default true?

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py [hunk 6]
```diff
@@ -28,25 +31,28 @@ class KFHungarianTracker(Node):
...
+                ("infer_orientation_from_velocity", False),
```
Comment (SteveMacenski) [inline kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: Default true?

File: nav2_dynamic_msgs/msg/Obstacle.msg (hunk 3 above)
Comment (SteveMacenski) [reply nav2_dynamic_msgs/msg/Obstacle.msg]: Yeah for both covars so that its in line with the same in other sensor message types

**Comment by SteveMacenski** [comment]
Otherwise LGTM!

**Comment by SteveMacenski** [comment]
Dont you need to parse the matrix correctly now? 0/4/8?

File: nav2_dynamic_msgs/msg/Obstacle.msg (hunk 3 above)
Comment (alexanderjyuen) [reply nav2_dynamic_msgs/msg/Obstacle.msg]: fixed in 5af5d49300af9cd023721984c2fae560ae49daee

File: kf_hungarian_tracker/kf_hungarian_tracker/obstacle_class.py (hunk 4 above)
Comment (alexanderjyuen) [reply kf_hungarian_tracker/kf_hungarian_tracker/obstacle_class.py]: fixed in 3299d8e7a0fef70b9dd6a77d7050eb4e61aed4ab

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py (hunk 5 above)
Comment (alexanderjyuen) [reply kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: fixed in 3299d8e7a0fef70b9dd6a77d7050eb4e61aed4ab

File: kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py (hunk 6 above)
Comment (alexanderjyuen) [reply kf_hungarian_tracker/kf_hungarian_tracker/kf_hungarian_node.py]: fixed in 3299d8e7a0fef70b9dd6a77d7050eb4e61aed4ab

**Comment by alexanderjyuen** [comment]
> Dont you need to parse the matrix correctly now? 0/4/8?

The current version of navigation2_dynamic does not us the measurement covariance in a meaningful way. It uses whatever the hardcoded measurement covariance, I could make the change in this PR or we could open another PR to do this

Right now the covariance from the radar just gets passed straight through

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
> I could make the change in this PR or we could open another PR to do this

Yeah, lets make that a follow up PR. LGTM, merging :+1:

**Pull request merged**